File: apps/api/src/lib/pagination.ts
import {
  BadRequestError,
  hashSha256,
  hmacSign,
  timingSafeEqual,
} from "@usevon/utils";
import { and, eq, gt, lt, or, type SQL, type SQLWrapper } from "drizzle-orm";
import { env } from "@/env";
import type { CursorPaginationQueryType } from "@/lib/models";

const CURSOR_VERSION = "v1";
const CURSOR_SIGNATURE_LENGTH = 24;
const CURSOR_SCOPE_HASH_LENGTH = 16;
const CURSOR_MAX_LENGTH = 256;
const UUID_REGEX =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
const CURSOR_REGEX =
  /^v1\.[0-9a-z]+\.[0-9a-f-]{36}\.[ad]\.[0-9a-f]{16}\.[0-9a-f]{24}$/i;
const SCOPE_HASH_REGEX = /^[0-9a-f]{16}$/i;
const CURSOR_SIGNATURE_REGEX = /^[0-9a-f]{24}$/i;

export const PAGINATION_ERROR_CODES = {
  INVALID_CURSOR: "Invalid cursor",
} as const;

export type CursorSort = "asc" | "desc";

export type CursorPageInput = {
  limit: number;
  cursor: string | null;
};

export type CursorPosition = {
  createdAt: Date;
  id: string;
};

type CursorInput = CursorPosition & {
  sort: CursorSort;
  scopeHash: string;
};

type CursorListQueryArgs<Row> = {
  pagination: CursorPageInput;
  sort: CursorSort;
  scope: unknown;
  createdAtColumn: SQLWrapper;
  idColumn: SQLWrapper;
  baseCondition: SQL<unknown>;
  fetchRows: (where: SQL<unknown>, limit: number) => Promise<Row[]>;
  toCursorPosition: (row: Row) => CursorPosition;
};

export const toCursorPageInput = (
  input: CursorPaginationQueryType
): CursorPageInput => ({
  limit: input.limit ?? 20,
  cursor: input.cursor ?? null,
});

const invalidCursor = (): never => {
  throw new BadRequestError(PAGINATION_ERROR_CODES.INVALID_CURSOR);
};

const sortToDirection = (sort: CursorSort): "a" | "d" =>
  sort === "asc" ? "a" : "d";

const directionToSort = (direction: string): CursorSort =>
  direction === "a" ? "asc" : "desc";

const signCursor = (unsigned: string): string =>
  hmacSign(unsigned, env.BETTER_AUTH_SECRET).slice(0, CURSOR_SIGNATURE_LENGTH);

export const buildCursorScopeHash = (scope: unknown): string =>
  hashSha256(JSON.stringify(scope)).slice(0, CURSOR_SCOPE_HASH_LENGTH);

export const encodeCursor = (input: CursorInput): string => {
  if (!UUID_REGEX.test(input.id)) {
    throw new Error("Cursor id must be a UUID");
  }
  if (!SCOPE_HASH_REGEX.test(input.scopeHash)) {
    throw new Error("Cursor scope hash must be 16 hex characters");
  }

  const createdAtMs = input.createdAt.getTime();
  if (!Number.isFinite(createdAtMs) || createdAtMs < 0) {
    throw new Error("Cursor createdAt is invalid");
  }

  const timestamp = Math.floor(createdAtMs).toString(36);
  const direction = sortToDirection(input.sort);
  const unsigned = `${CURSOR_VERSION}.${timestamp}.${input.id}.${direction}.${input.scopeHash}`;
  const signature = signCursor(unsigned);

  return `${unsigned}.${signature}`;
};

export const decodeCursor = (
  cursor: string | null,
  expected: { sort: CursorSort; scopeHash: string }
): CursorPosition | null => {
  if (!cursor) {
    return null;
  }

  if (cursor.length > CURSOR_MAX_LENGTH || !CURSOR_REGEX.test(cursor)) {
    invalidCursor();
  }

  const parts = cursor.split(".");
  if (parts.length !== 6) {
    invalidCursor();
  }

  const [version, timestamp, id, direction, scopeHash, signature] = parts;
  const hasAllParts = [
    version,
    timestamp,
    id,
    direction,
    scopeHash,
    signature,
  ].every(Boolean);
  if (!hasAllParts) {
    invalidCursor();
  }

  if (version !== CURSOR_VERSION || !UUID_REGEX.test(id)) {
    invalidCursor();
  }

  if (!SCOPE_HASH_REGEX.test(scopeHash)) {
    invalidCursor();
  }

  if (!CURSOR_SIGNATURE_REGEX.test(signature)) {
    invalidCursor();
  }

  const unsigned = `${version}.${timestamp}.${id}.${direction}.${scopeHash}`;
  const expectedSignature = signCursor(unsigned);
  if (!timingSafeEqual(expectedSignature, signature)) {
    invalidCursor();
  }

  if (directionToSort(direction) !== expected.sort) {
    invalidCursor();
  }

  if (scopeHash !== expected.scopeHash) {
    invalidCursor();
  }

  const createdAtMs = Number.parseInt(timestamp, 36);
  if (!Number.isFinite(createdAtMs) || createdAtMs < 0) {
    invalidCursor();
  }

  const createdAt = new Date(createdAtMs);
  if (Number.isNaN(createdAt.getTime())) {
    invalidCursor();
  }

  return { createdAt, id };
};

export const buildCursorCondition = (
  createdAtColumn: SQLWrapper,
  idColumn: SQLWrapper,
  cursor: CursorPosition,
  sort: CursorSort
): SQL<unknown> => {
  const createdAtCompare =
    sort === "asc"
      ? gt(createdAtColumn, cursor.createdAt)
      : lt(createdAtColumn, cursor.createdAt);
  const idCompare =
    sort === "asc" ? gt(idColumn, cursor.id) : lt(idColumn, cursor.id);

  return or(
    createdAtCompare,
    and(eq(createdAtColumn, cursor.createdAt), idCompare)
  ) as SQL<unknown>;
};

export const sliceCursorPage = <T>(rows: T[], limit: number) => {
  const hasMore = rows.length > limit;
  const items = hasMore ? rows.slice(0, limit) : rows;
  return {
    items,
    hasMore,
    lastItem: items.at(-1) ?? null,
  };
};

export const runCursorListQuery = async <Row>(
  args: CursorListQueryArgs<Row>
): Promise<{ items: Row[]; nextCursor: string | null }> => {
  const scopeHash = buildCursorScopeHash(args.scope);
  const cursor = decodeCursor(args.pagination.cursor, {
    sort: args.sort,
    scopeHash,
  });

  const where = cursor
    ? (and(
        args.baseCondition,
        buildCursorCondition(
          args.createdAtColumn,
          args.idColumn,
          cursor,
          args.sort
        )
      ) ?? args.baseCondition)
    : args.baseCondition;

  const rows = await args.fetchRows(where, args.pagination.limit + 1);
  const { items, hasMore, lastItem } = sliceCursorPage(
    rows,
    args.pagination.limit
  );

  return {
    items,
    nextCursor:
      hasMore && lastItem
        ? encodeCursor({
            ...args.toCursorPosition(lastItem),
            sort: args.sort,
            scopeHash,
          })
        : null,
  };
};

File: services/crates/von-api/src/pagination.rs
use chrono::{DateTime, TimeZone, Utc};
use hmac::{Hmac, Mac};
use serde::Deserialize;
use sha2::{Digest, Sha256};
use sqlx::postgres::PgRow;
use sqlx::{PgPool, Row};
use std::sync::OnceLock;
use von_error::{Error, Result};

const CURSOR_VERSION: &str = "v1";
const SIGNATURE_LENGTH: usize = 24;
const SCOPE_HASH_LENGTH: usize = 16;
const MAX_LENGTH: usize = 256;
const DEFAULT_LIMIT: i64 = 20;
const MAX_LIMIT: i64 = 100;

static SIGNING_SECRET: OnceLock<Option<String>> = OnceLock::new();

fn signing_secret() -> Result<&'static str> {
    SIGNING_SECRET
        .get_or_init(|| std::env::var("BETTER_AUTH_SECRET").ok())
        .as_deref()
        .ok_or_else(|| Error::Configuration("BETTER_AUTH_SECRET is required".to_owned()))
}

#[derive(Deserialize, utoipa::IntoParams)]
#[into_params(parameter_in = Query)]
pub struct PaginationQuery {
    /// Maximum items to return, defaults to 20.
    #[param(minimum = 1, maximum = 100, example = 20)]
    pub limit: Option<i64>,
    /// Opaque cursor from a previous response's nextCursor.
    pub cursor: Option<String>,
}

pub fn clamp_limit(limit: Option<i64>) -> i64 {
    limit.unwrap_or(DEFAULT_LIMIT).clamp(1, MAX_LIMIT)
}

pub struct CursorPosition {
    pub created_at: DateTime<Utc>,
    pub id: String,
}

#[derive(Clone, Copy, PartialEq, Eq, Debug, Default, Deserialize, utoipa::ToSchema)]
#[serde(rename_all = "lowercase")]
pub enum CursorSort {
    Asc,
    #[default]
    Desc,
}

impl CursorSort {
    fn direction(self) -> &'static str {
        match self {
            Self::Asc => "a",
            Self::Desc => "d",
        }
    }

    pub fn compare(self) -> &'static str {
        match self {
            Self::Asc => ">",
            Self::Desc => "<",
        }
    }

    pub fn order(self) -> &'static str {
        match self {
            Self::Asc => "ASC",
            Self::Desc => "DESC",
        }
    }

    pub fn as_str(self) -> &'static str {
        match self {
            Self::Asc => "asc",
            Self::Desc => "desc",
        }
    }
}

fn to_base36(mut value: u64) -> String {
    const ALPHABET: &[u8] = b"0123456789abcdefghijklmnopqrstuvwxyz";
    if value == 0 {
        return "0".to_owned();
    }
    let mut out = Vec::new();
    while value > 0 {
        out.push(ALPHABET[(value % 36) as usize]);
        value /= 36;
    }
    out.reverse();
    String::from_utf8(out).unwrap_or_default()
}

fn from_base36(value: &str) -> Option<i64> {
    let mut out: i64 = 0;
    for byte in value.bytes() {
        let digit = match byte {
            b'0'..=b'9' => i64::from(byte - b'0'),
            b'a'..=b'z' => i64::from(byte - b'a') + 10,
            b'A'..=b'Z' => i64::from(byte - b'A') + 10,
            _ => return None,
        };
        out = out.checked_mul(36)?.checked_add(digit)?;
    }
    Some(out)
}

fn sign(unsigned: &str) -> Result<String> {
    let mut mac = Hmac::<Sha256>::new_from_slice(signing_secret()?.as_bytes())
        .map_err(|_| Error::Configuration("invalid cursor signing secret".to_owned()))?;
    mac.update(unsigned.as_bytes());
    let mut hex = hex::encode(mac.finalize().into_bytes());
    hex.truncate(SIGNATURE_LENGTH);
    Ok(hex)
}

/// Must match the TypeScript scope hash byte for byte, which hashes the JSON
/// object with its keys in declaration order rather than sorted.
pub fn scope_hash_json(json: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(json.as_bytes());
    let mut hex = hex::encode(hasher.finalize());
    hex.truncate(SCOPE_HASH_LENGTH);
    hex
}

fn scope_hash(resource: &str, organization_id: &str) -> String {
    scope_hash_json(&format!(
        "{{\"resource\":\"{resource}\",\"organizationId\":\"{organization_id}\"}}"
    ))
}

fn is_uuid(value: &str) -> bool {
    uuid::Uuid::parse_str(value).is_ok() && value.len() == 36
}

pub fn encode_cursor_sorted(
    position: &CursorPosition,
    scope: &str,
    sort: CursorSort,
) -> Result<String> {
    let millis = position.created_at.timestamp_millis();
    if millis < 0 {
        return Err(Error::BadRequest("Invalid cursor".to_owned()));
    }
    let unsigned = format!(
        "{CURSOR_VERSION}.{}.{}.{}.{scope}",
        to_base36(millis as u64),
        position.id,
        sort.direction()
    );
    let signature = sign(&unsigned)?;
    Ok(format!("{unsigned}.{signature}"))
}

pub fn decode_cursor_sorted(
    cursor: Option<&str>,
    scope: &str,
    sort: CursorSort,
) -> Result<Option<CursorPosition>> {
    let Some(cursor) = cursor.filter(|c| !c.is_empty()) else {
        return Ok(None);
    };

    let invalid = || Error::BadRequest("Invalid cursor".to_owned());

    if cursor.len() > MAX_LENGTH {
        return Err(invalid());
    }

    let parts: Vec<&str> = cursor.split('.').collect();
    let [version, timestamp, id, direction, cursor_scope, signature] = parts[..] else {
        return Err(invalid());
    };

    if version != CURSOR_VERSION || !is_uuid(id) || direction != sort.direction() {
        return Err(invalid());
    }
    if cursor_scope.len() != SCOPE_HASH_LENGTH
        || !cursor_scope.bytes().all(|b| b.is_ascii_hexdigit())
        || signature.len() != SIGNATURE_LENGTH
        || !signature.bytes().all(|b| b.is_ascii_hexdigit())
    {
        return Err(invalid());
    }

    let unsigned = format!("{version}.{timestamp}.{id}.{direction}.{cursor_scope}");
    let expected = sign(&unsigned)?;
    let matches = expected.len() == signature.len()
        && expected
            .bytes()
            .zip(signature.bytes())
            .fold(0u8, |acc, (a, b)| acc | (a ^ b))
            == 0;
    if !matches || cursor_scope != scope {
        return Err(invalid());
    }

    let millis = from_base36(timestamp).ok_or_else(invalid)?;
    let created_at = Utc
        .timestamp_millis_opt(millis)
        .single()
        .ok_or_else(invalid)?;

    Ok(Some(CursorPosition {
        created_at,
        id: id.to_owned(),
    }))
}

pub async fn find_org_row(
    pool: &PgPool,
    table: &str,
    columns: &str,
    organization_id: &str,
    id: &str,
) -> Result<Option<PgRow>> {
    let Ok(uuid) = uuid::Uuid::parse_str(id) else {
        return Ok(None);
    };
    let row = sqlx::query(&format!(
        "SELECT {columns} FROM {table} WHERE id = $1 AND organization_id = $2::uuid LIMIT 1"
    ))
    .bind(uuid)
    .bind(organization_id)
    .fetch_optional(pool)
    .await?;
    Ok(row)
}

pub async fn fetch_org_page(
    pool: &PgPool,
    table: &str,
    columns: &str,
    resource: &str,
    organization_id: &str,
    pagination: &PaginationQuery,
) -> Result<(Vec<PgRow>, Option<String>)> {
    let scope = scope_hash(resource, organization_id);
    let cursor = decode_cursor_sorted(pagination.cursor.as_deref(), &scope, CursorSort::Desc)?;
    let limit = clamp_limit(pagination.limit);

    // One extra row tells us whether another page exists without a second query.
    let mut sql = format!("SELECT {columns} FROM {table} WHERE organization_id = $1::uuid");
    if cursor.is_some() {
        sql.push_str(" AND (created_at < $3 OR (created_at = $3 AND id < $4))");
    }
    sql.push_str(" ORDER BY created_at DESC, id DESC LIMIT $2");

    let mut query = sqlx::query(&sql).bind(organization_id).bind(limit + 1);
    if let Some(position) = &cursor {
        let id = uuid::Uuid::parse_str(&position.id)
            .map_err(|_| Error::BadRequest("Invalid cursor".to_owned()))?;
        query = query.bind(position.created_at).bind(id);
    }

    let mut rows = query.fetch_all(pool).await?;

    let has_more = rows.len() > limit as usize;
    rows.truncate(limit as usize);

    let next_cursor = match rows.last().filter(|_| has_more) {
        Some(last) => Some(encode_cursor_sorted(
            &CursorPosition {
                created_at: last.try_get("created_at")?,
                id: last.try_get("id")?,
            },
            &scope,
            CursorSort::Desc,
        )?),
        None => None,
    };

    Ok((rows, next_cursor))
}

/// Pairs must stay in the order the TypeScript caller wrote them or the hash will not match JSON.stringify.
pub fn scope_hash_fields(fields: &[(&str, serde_json::Value)]) -> String {
    let mut json = String::from("{");
    for (index, (key, value)) in fields.iter().enumerate() {
        if index > 0 {
            json.push(',');
        }
        json.push_str(&serde_json::Value::String((*key).to_owned()).to_string());
        json.push(':');
        json.push_str(&value.to_string());
    }
    json.push('}');
    scope_hash_json(&json)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn base36_matches_javascript() {
        assert_eq!(to_base36(0), "0");
        assert_eq!(to_base36(35), "z");
        assert_eq!(to_base36(1_700_000_000_000), "loyw3v28");
        assert_eq!(from_base36("loyw3v28"), Some(1_700_000_000_000));
    }

    #[test]
    fn round_trips_a_cursor() {
        unsafe { std::env::set_var("BETTER_AUTH_SECRET", "test-secret") };
        let scope = scope_hash("endpoints", "org-1");
        let position = CursorPosition {
            created_at: Utc.timestamp_millis_opt(1_700_000_000_000).unwrap(),
            id: "3f7c1a2e-5b6d-4e8f-9a0b-1c2d3e4f5a6b".to_owned(),
        };
        let encoded = encode_cursor_sorted(&position, &scope, CursorSort::Desc).expect("encode");
        let decoded = decode_cursor_sorted(Some(&encoded), &scope, CursorSort::Desc)
            .expect("decode")
            .expect("some");
        assert_eq!(decoded.id, position.id);
        assert_eq!(decoded.created_at, position.created_at);
    }

    /// The expected values come from running JSON.stringify plus sha256 in bun,
    /// so a change to the field order here fails rather than silently rejecting
    /// every cursor the other service issued.
    #[test]
    fn scope_hash_matches_javascript() {
        use serde_json::json;

        assert_eq!(
            scope_hash_fields(&[
                ("resource", json!("webhook-events")),
                ("organizationId", json!("org-1")),
                ("eventTypes", json!(None::<Vec<String>>)),
                ("from", json!(null)),
                ("to", json!(null)),
                ("sort", json!("desc")),
            ]),
            "fdabbf55c0fc9325"
        );
        assert_eq!(
            scope_hash_fields(&[
                ("resource", json!("webhook-events")),
                ("organizationId", json!("org-1")),
                ("eventTypes", json!(["a.b", "c.d"])),
                ("from", json!("2024-01-01T00:00:00.000Z")),
                ("to", json!(null)),
                ("sort", json!("asc")),
            ]),
            "aa8035901c10ed52"
        );
        assert_eq!(
            scope_hash_fields(&[
                ("resource", json!("webhook-deliveries")),
                ("organizationId", json!("org-1")),
                ("eventId", json!("e-1")),
                ("status", json!(null)),
                ("endpointId", json!(null)),
                ("from", json!(null)),
                ("to", json!(null)),
                ("sort", json!("desc")),
            ]),
            "bf7e755f828eebb7"
        );
        assert_eq!(
            scope_hash_fields(&[
                ("resource", json!("webhook-delivery-attempts")),
                ("organizationId", json!("org-1")),
                ("deliveryId", json!("d-1")),
                ("sort", json!("asc")),
            ]),
            "51d23ebdff40792f"
        );
    }

    #[test]
    fn ascending_and_descending_cursors_do_not_interchange() {
        unsafe { std::env::set_var("BETTER_AUTH_SECRET", "test-secret") };
        let scope = scope_hash("webhook-events", "org-1");
        let position = CursorPosition {
            created_at: Utc.timestamp_millis_opt(1_700_000_000_000).unwrap(),
            id: "3f7c1a2e-5b6d-4e8f-9a0b-1c2d3e4f5a6b".to_owned(),
        };
        let ascending = encode_cursor_sorted(&position, &scope, CursorSort::Asc).expect("encode");
        assert!(decode_cursor_sorted(Some(&ascending), &scope, CursorSort::Asc).is_ok());
        assert!(decode_cursor_sorted(Some(&ascending), &scope, CursorSort::Desc).is_err());
    }

    #[test]
    fn rejects_a_cursor_from_another_scope() {
        unsafe { std::env::set_var("BETTER_AUTH_SECRET", "test-secret") };
        let position = CursorPosition {
            created_at: Utc.timestamp_millis_opt(1_700_000_000_000).unwrap(),
            id: "3f7c1a2e-5b6d-4e8f-9a0b-1c2d3e4f5a6b".to_owned(),
        };
        let encoded = encode_cursor_sorted(
            &position,
            &scope_hash("endpoints", "org-1"),
            CursorSort::Desc,
        )
        .expect("encode");
        assert!(
            decode_cursor_sorted(
                Some(&encoded),
                &scope_hash("endpoints", "org-2"),
                CursorSort::Desc
            )
            .is_err()
        );
    }
}
